File: java/greedy/intro.java
/**
 * Author: Ravi Ranjan
 * GitHub: @its-ravi-ranjan
 * Email: contact70raviranjan@gmail.com
 * Created: 2024
 * Purpose: Interview preparation and greedy algorithm learning guide
 * 
 * ========================================
 * GREEDY ALGORITHMS - COMPLETE GUIDE
 * ========================================
 * 
 * What is a Greedy Algorithm?
 * -----------------------------
 * A greedy algorithm makes locally optimal choices at each step to find a global optimum.
 * It never looks back or changes previous decisions.
 * 
 * Key Principles:
 * ---------------
 * 1. Greedy Choice Property: Locally optimal choice leads to globally optimal solution
 * 2. Optimal Substructure: Optimal solution contains optimal sub-solutions
 * 3. No Backtracking: Once a choice is made, it's never reconsidered
 * 4. Simple and Fast: Usually O(n log n) or O(n) time complexity
 * 
 * Visual Example:
 * 
 * Activity Selection: Choose maximum non-overlapping activities
 * 
 * Activities: [1,4], [2,6], [5,7], [3,8], [5,9]
 * Sorted by end time: [1,4], [2,6], [5,7], [3,8], [5,9]
 * Greedy choice: Always pick activity with earliest end time
 * Result: [1,4], [5,7] - 2 activities
 * 
 * ========================================
 * WHEN TO RECOGNIZE GREEDY PROBLEMS
 * ========================================
 * 
 * 🔹 5. Ninja Techniques to Recognize Greedy Problems
 * 
 * ✅ "Activity selection" — pick optimal activities
 *    Examples: Meeting rooms, job scheduling, task assignment
 *    Pattern: Need to select maximum non-overlapping activities
 * 
 * ✅ "Fractional knapsack" — take items partially
 *    Examples: Loading truck, resource allocation, profit maximization
 *    Pattern: Can take fractional amounts of items
 * 
 * ✅ "Huffman coding" — build optimal prefix codes
 *    Examples: Data compression, file encoding, character frequency
 *    Pattern: Need to build optimal binary tree for encoding
 * 
 * ✅ "Minimum spanning tree" — connect all nodes
 *    Examples: Network wiring, road construction, cluster analysis
 *    Pattern: Need to connect all nodes with minimum cost
 * 
 * ✅ "Coin change" — minimum coins for amount
 *    Examples: Vending machine, cash register, payment systems
 *    Pattern: Need minimum number of coins for given amount
 * 
 * ✅ "Interval scheduling" — schedule events optimally
 *    Examples: Class scheduling, meeting rooms, resource booking
 *    Pattern: Need to schedule maximum events without conflicts
 * 
 * ========================================
 * GREEDY ALGORITHM TYPES
 * ========================================
 * 
 * 1. SELECTION PROBLEMS
 * ----------------------
 * Use: Choose optimal subset from given options
 * Strategy: Sort and select based on criteria
 * Time: O(n log n) - sorting
 * Space: O(1) - in-place
 * 
 * Examples:
 * - Activity Selection: Sort by end time, pick earliest ending
 * - Job Scheduling: Sort by deadline, schedule earliest deadline
 * - Meeting Rooms: Sort by start time, assign to available room
 * 
 * 2. OPTIMIZATION PROBLEMS
 * -------------------------
 * Use: Find optimal value (min/max) with constraints
 * Strategy: Make locally optimal choice at each step
 * Time: O(n log n) - usually involves sorting
 * Space: O(1) - constant extra space
 * 
 * Examples:
 * - Fractional Knapsack: Sort by value/weight ratio
 * - Minimum Coins: Use largest denomination first
 * - Maximum Profit: Sort by profit, take highest
 * 
 * 3. CONSTRUCTION PROBLEMS
 * -------------------------
 * Use: Build optimal structure step by step
 * Strategy: Add elements in optimal order
 * Time: O(n log n) - building and maintaining structure
 * Space: O(n) - store the structure
 * 
 * Examples:
 * - Huffman Coding: Build tree from frequency
 * - Minimum Spanning Tree: Add edges in order of weight
 * - Optimal Merge: Merge smallest files first
 * 
 * ========================================
 * COMMON PROBLEM PATTERNS
 * ========================================
 * 
 * Pattern 1: "Activity Selection"
 * --------------------------------
 * Keywords: "activities", "meetings", "events", "non-overlapping"
 * Solution: Sort by end time, pick earliest ending
 * 
 * Pattern 2: "Fractional Knapsack"
 * ----------------------------------
 * Keywords: "fractional", "partial", "ratio", "value per unit"
 * Solution: Sort by value/weight ratio, take highest first
 * 
 * Pattern 3: "Coin Change"
 * -------------------------
 * Keywords: "minimum coins", "change", "denomination"
 * Solution: Use largest denomination first
 * 
 * Pattern 4: "Interval Scheduling"
 * ---------------------------------
 * Keywords: "intervals", "schedule", "conflicts", "overlap"
 * Solution: Sort by end time, select non-overlapping
 * 
 * Pattern 5: "Optimal Merge"
 * ---------------------------
 * Keywords: "merge", "combine", "cost", "minimum cost"
 * Solution: Always merge smallest elements first
 * 
 * Pattern 6: "Resource Allocation"
 * ---------------------------------
 * Keywords: "allocate", "assign", "distribute", "efficient"
 * Solution: Sort by efficiency, allocate to best option
 * 
 * ========================================
 * IMPLEMENTATION STRATEGIES
 * ========================================
 * 
 * Java Greedy Implementation:
 * ----------------------------
 * 
 * // Activity Selection
 * public int maxActivities(int[][] activities) {
 *     // Sort by end time
 *     Arrays.sort(activities, (a, b) -> a[1] - b[1]);
 *     
 *     int count = 0;
 *     int lastEnd = 0;
 *     
 *     for (int[] activity : activities) {
 *         if (activity[0] >= lastEnd) {
 *             count++;
 *             lastEnd = activity[1];
 *         }
 *     }
 *     return count;
 * }
 * 
 * // Fractional Knapsack
 * public double fractionalKnapsack(int[] weights, int[] values, int capacity) {
 *     int n = weights.length;
 *     double[][] items = new double[n][3]; // [index, value/weight, weight]
 *     
 *     for (int i = 0; i < n; i++) {
 *         items[i][0] = i;
 *         items[i][1] = (double) values[i] / weights[i];
 *         items[i][2] = weights[i];
 *     }
 *     
 *     // Sort by value/weight ratio (descending)
 *     Arrays.sort(items, (a, b) -> Double.compare(b[1], a[1]));
 *     
 *     double totalValue = 0;
 *     int remainingCapacity = capacity;
 *     
 *     for (double[] item : items) {
 *         if (remainingCapacity >= item[2]) {
 *             totalValue += values[(int)item[0]];
 *             remainingCapacity -= item[2];
 *         } else {
 *             totalValue += item[1] * remainingCapacity;
 *             break;
 *         }
 *     }
 *     return totalValue;
 * }
 * 
 * // Minimum Coins
 * public int minCoins(int[] coins, int amount) {
 *     Arrays.sort(coins); // Sort in descending order
 *     int count = 0;
 *     
 *     for (int i = coins.length - 1; i >= 0; i--) {
 *         while (amount >= coins[i]) {
 *             amount -= coins[i];
 *             count++;
 *         }
 *     }
 *     return amount == 0 ? count : -1;
 * }
 * 
 * ========================================
 * TIME COMPLEXITY GUIDE
 * ========================================
 * 
 * Problem Type        | Time Complexity | Space Complexity
 * --------------------|-----------------|------------------
 * Activity Selection  | O(n log n)     | O(1)
 * Fractional Knapsack | O(n log n)     | O(n)
 * Coin Change         | O(n log n)     | O(1)
 * Interval Scheduling | O(n log n)     | O(1)
 * Huffman Coding      | O(n log n)     | O(n)
 * MST (Kruskal)       | O(E log E)     | O(V)
 * 
 * Optimization        | Time            | Space
 * --------------------|-----------------|------------------
 * In-place sorting    | O(n log n)      | O(1)
 * Heap-based          | O(n log n)      | O(n)
 * Linear scan         | O(n)            | O(1)
 * 
 * ========================================
 * INTERVIEW TIPS
 * ========================================
 * 
 * 1. RECOGNITION TIPS:
 *    - Look for "maximum", "minimum", "optimal" keywords
 *    - Check if problem involves selection or optimization
 *    - Identify if local optimal choice leads to global optimal
 *    - See if problem can be solved by sorting and selecting
 * 
 * 2. IMPLEMENTATION TIPS:
 *    - Start by sorting the input
 *    - Make greedy choice at each step
 *    - Don't look back or change previous decisions
 *    - Consider edge cases and constraints
 * 
 * 3. OPTIMIZATION TIPS:
 *    - Use in-place sorting when possible
 *    - Consider using heaps for dynamic selection
 *    - Think about early termination conditions
 *    - Optimize space usage
 * 
 * 4. COMMON MISTAKES:
 *    - Not sorting input properly
 *    - Wrong sorting criteria
 *    - Not handling edge cases
 *    - Forgetting to update state after choice
 *    - Not considering all constraints
 * 
 * ========================================
 * PRACTICE PROBLEMS BY DIFFICULTY
 * ========================================
 * 
 * EASY:
 * - Assign Cookies
 * - Lemonade Change
 * - Best Time to Buy/Sell Stock
 * - Can Place Flowers
 * 
 * MEDIUM:
 * - Jump Game
 * - Gas Station
 * - Task Scheduler
 * - Partition Labels
 * 
 * HARD:
 * - Candy
 * - Remove K Digits
 * - Create Maximum Number
 * - Minimum Number of Arrows
 * 
 * ========================================
 * QUICK REFERENCE
 * ========================================
 * 
 * When to use Greedy:
 * - Problem has optimal substructure
 * - Greedy choice property holds
 * - Need to make locally optimal choices
 * - Problem involves selection or optimization
 * - Can be solved by sorting and selecting
 * 
 * Which Strategy to use:
 * - Selection: Sort and pick based on criteria
 * - Optimization: Make locally optimal choice
 * - Construction: Build step by step optimally
 * 
 * Key Steps:
 * 1. Identify the greedy choice property
 * 2. Sort input based on criteria
 * 3. Make greedy choice at each step
 * 4. Update state and continue
 * 5. Handle edge cases
 * 
 * Time Complexity:
 * - Usually O(n log n) due to sorting
 * - Sometimes O(n) for linear scan
 * - Space usually O(1) or O(n)
 * 
 * This guide should help you quickly identify greedy problems and choose the right approach!
 */

public class intro {
    
    /**
     * This is a reference guide for greedy algorithm problems.
     * No implementation needed - just concepts and patterns.
     */
    
    public static void main(String[] args) {
        System.out.println("=== GREEDY ALGORITHMS GUIDE ===");
        System.out.println("Use this as a quick reference for greedy problems.");
        System.out.println("Key patterns to recognize:");
        System.out.println("1. 'Activity selection' problems");
        System.out.println("2. 'Fractional knapsack' problems");
        System.out.println("3. 'Huffman coding' problems");
        System.out.println("4. 'Minimum spanning tree' problems");
        System.out.println("5. 'Coin change' problems");
        System.out.println("6. 'Interval scheduling' problems");
        System.out.println();
        System.out.println("Choose strategy based on problem:");
        System.out.println("- Selection: Sort and pick based on criteria");
        System.out.println("- Optimization: Make locally optimal choice");
        System.out.println("- Construction: Build step by step optimally");
    }
} 

File: javascript/heap/intro.js
/**
 * Author: Ravi Ranjan
 * GitHub: @its-ravi-ranjan
 * Email: contact70raviranjan@gmail.com
 * Created: 2024
 * Purpose: Interview preparation and heap learning guide
 * 
 * ========================================
 * HEAP DATA STRUCTURE - COMPLETE GUIDE
 * ========================================
 * 
 * What is a Heap?
 * ----------------
 * A heap is a complete binary tree (or array) that satisfies the heap property:
 * 
 * Min Heap: Parent ≤ children → smallest element at root
 * Max Heap: Parent ≥ children → largest element at root
 * 
 * Visual Representation:
 * 
 * Min Heap:                    Max Heap:
 *       1                          9
 *      / \                       / \
 *     3   5                     7   8
 *    / \ / \                   / \ / \
 *   4  6 7  8                 5  6 3  4
 * 
 * Array Representation:
 * Min Heap: [1, 3, 5, 4, 6, 7, 8]
 * Max Heap: [9, 7, 8, 5, 6, 3, 4]
 * 
 * Key Properties:
 * --------------
 * 1. Complete Binary Tree: All levels filled except possibly the last
 * 2. Heap Property: Parent-child relationship maintained
 * 3. Root Access: O(1) access to min/max element
 * 4. Insert/Delete: O(log n) operations
 * 5. Build Heap: O(n) from array
 * 
 * Array Indexing:
 * --------------
 * For element at index i:
 * - Left Child: 2*i + 1
 * - Right Child: 2*i + 2
 * - Parent: Math.floor((i-1)/2)
 * 
 * Basic Operations:
 * ----------------
 * 1. Insert: Add to end, bubble up
 * 2. Delete: Replace root with last, bubble down
 * 3. Peek: Return root element
 * 4. Heapify: Convert array to heap
 * 
 * ========================================
 * WHEN TO RECOGNIZE HEAP PROBLEMS
 * ========================================
 * 
 * 🔹 4. Ninja Techniques to Recognize Heap Problems
 * 
 * ✅ "Top K…" — almost always heap
 *    Examples: Top K frequent elements, Top K largest numbers
 *    Pattern: Need to maintain K elements with specific property
 * 
 * ✅ "Kth largest/smallest" — use heap of size K
 *    Examples: Kth largest element in array, Kth smallest in stream
 *    Pattern: Need to track K elements, heap size = K
 * 
 * ✅ "Stream" or "real-time median" — use two heaps
 *    Examples: Find median of data stream, sliding window median
 *    Pattern: Need to track middle element(s) in dynamic data
 * 
 * ✅ "Scheduling with priority" — custom PriorityQueue
 *    Examples: Task scheduler, meeting rooms, CPU scheduling
 *    Pattern: Elements have priorities, need to process by priority
 * 
 * ✅ "Frequent elements" — count → put in heap
 *    Examples: Top K frequent words, most common elements
 *    Pattern: Count frequencies, then heap for top K
 * 
 * ✅ "Merge sorted arrays/lists" — min heap of head elements
 *    Examples: Merge K sorted lists, merge K sorted arrays
 *    Pattern: Multiple sorted sources, need to merge efficiently
 * 
 * ========================================
 * MIN HEAP vs MAX HEAP USE CASES
 * ========================================
 * 
 * Use Case                    | Use Min Heap | Use Max Heap
 * ----------------------------|--------------|--------------
 * Find K smallest            | ✅           | ❌
 * Find K largest             | ❌           | ✅
 * Running Median             | ✅ (high)    | ✅ (low)
 * Task scheduling            | ✅           | ❌
 * Custom rules (freq, age)   | ✅/custom    | ✅/custom
 * 
 * Detailed Use Cases:
 * -------------------
 * 
 * 1. FIND K SMALLEST ELEMENTS
 *    - Use: Min Heap of size K
 *    - Logic: Keep K smallest elements, replace if smaller found
 *    - Example: [3,1,5,2,4], K=3 → [1,2,3]
 * 
 * 2. FIND K LARGEST ELEMENTS
 *    - Use: Max Heap of size K
 *    - Logic: Keep K largest elements, replace if larger found
 *    - Example: [3,1,5,2,4], K=3 → [5,4,3]
 * 
 * 3. RUNNING MEDIAN
 *    - Use: Two heaps (min + max)
 *    - Logic: Min heap for upper half, max heap for lower half
 *    - Example: Stream [1,2,3,4] → median = 2.5
 * 
 * 4. TASK SCHEDULING
 *    - Use: Min Heap (by end time)
 *    - Logic: Process tasks with earliest deadline first
 *    - Example: Meeting rooms, CPU scheduling
 * 
 * 5. FREQUENCY-BASED PROBLEMS
 *    - Use: Max Heap (by frequency)
 *    - Logic: Count frequencies, heap by count
 *    - Example: Top K frequent words
 * 
 * ========================================
 * COMMON PROBLEM PATTERNS
 * ========================================
 * 
 * Pattern 1: "Top K" Problems
 * ----------------------------
 * Keywords: "top", "most", "frequent", "largest", "smallest"
 * Solution: Maintain heap of size K
 * 
 * Pattern 2: "Stream" Problems
 * ----------------------------
 * Keywords: "stream", "real-time", "online", "continuous"
 * Solution: Two heaps for median, single heap for min/max
 * 
 * Pattern 3: "Scheduling" Problems
 * --------------------------------
 * Keywords: "schedule", "meeting", "room", "task", "priority"
 * Solution: Priority queue based on time/priority
 * 
 * Pattern 4: "Merge" Problems
 * ----------------------------
 * Keywords: "merge", "combine", "multiple sorted"
 * Solution: Min heap of head elements from each source
 * 
 * Pattern 5: "Kth Element" Problems
 * ---------------------------------
 * Keywords: "Kth", "nth", "rank", "order"
 * Solution: Heap of size K or quick select
 * 
 * ========================================
 * IMPLEMENTATION STRATEGIES
 * ========================================
 * 
 * JavaScript Heap Implementation Options:
 * ---------------------------------------
 * 
 * Option 1: Array-based with sort (Simple but inefficient)
 * const minHeap = [];
 * minHeap.push(value);
 * minHeap.sort((a, b) => a - b); // O(n log n) per operation
 * 
 * Option 2: Custom MinHeap Class (Recommended)
 * class MinHeap {
 *   constructor() { this.heap = []; }
 *   push(value) { // bubble up logic }
 *   pop() { // bubble down logic }
 *   peek() { return this.heap[0]; }
 * }
 * 
 * Option 3: Using Priority Queue Library
 * // If using a library like 'priority-queue'
 * const PriorityQueue = require('priority-queue');
 * const pq = new PriorityQueue({comparator: (a, b) => a - b});
 * 
 * Common Operations:
 * -----------------
 * heap.push(element);    // Insert
 * heap.pop();           // Remove and return root
 * heap[0];              // View root without removing
 * heap.length;          // Number of elements
 * heap.length === 0;    // Check if empty
 * 
 * ========================================
 * TIME COMPLEXITY GUIDE
 * ========================================
 * 
 * Operation          | Time Complexity | Space Complexity
 * -------------------|-----------------|------------------
 * Insert            | O(log n)        | O(1)
 * Delete            | O(log n)        | O(1)
 * Peek              | O(1)            | O(1)
 * Build Heap        | O(n)            | O(1)
 * Heap Sort         | O(n log n)      | O(1)
 * 
 * Note: Array-based with sort is O(n log n) per operation
 * 
 * ========================================
 * INTERVIEW TIPS
 * ========================================
 * 
 * 1. RECOGNITION TIPS:
 *    - Look for "K" in problem statement
 *    - Check for "top", "most", "frequent" keywords
 *    - Identify if it's a stream/online problem
 *    - See if multiple sorted sources need merging
 * 
 * 2. IMPLEMENTATION TIPS:
 *    - Always specify heap type (min/max)
 *    - Use appropriate comparator for custom objects
 *    - Consider heap size (K vs all elements)
 *    - Think about when to insert/remove
 *    - Prefer custom heap class over array.sort()
 * 
 * 3. OPTIMIZATION TIPS:
 *    - Use heap size K instead of all elements when possible
 *    - Consider two heaps for median problems
 *    - Use custom comparators for complex ordering
 *    - Think about early termination conditions
 *    - Avoid array.sort() in loops (O(n log n) each time)
 * 
 * 4. COMMON MISTAKES:
 *    - Forgetting to specify heap type (min vs max)
 *    - Using wrong heap size (K vs n)
 *    - Not handling edge cases (empty heap)
 *    - Incorrect comparator logic
 *    - Using array.sort() repeatedly instead of proper heap
 * 
 * ========================================
 * PRACTICE PROBLEMS BY DIFFICULTY
 * ========================================
 * 
 * EASY:
 * - Kth Largest Element in Array
 * - Top K Frequent Elements
 * - Meeting Rooms II
 * 
 * MEDIUM:
 * - Merge K Sorted Lists
 * - Find Median from Data Stream
 * - Task Scheduler
 * 
 * HARD:
 * - Sliding Window Median
 * - Merge K Sorted Arrays
 * - Design Twitter (feed system)
 * 
 * ========================================
 * QUICK REFERENCE
 * ========================================
 * 
 * When to use Heap:
 * - Need to track min/max elements
 * - Processing elements by priority
 * - Finding K largest/smallest
 * - Real-time median calculation
 * - Merging multiple sorted sources
 * 
 * Which Heap to use:
 * - Min Heap: K smallest, scheduling by end time
 * - Max Heap: K largest, frequency-based problems
 * - Two Heaps: Median, balanced partitioning
 * 
 * Key Operations:
 * - push(): Insert element
 * - pop(): Remove and return root
 * - peek(): View root (heap[0])
 * - length: Number of elements
 * 
 * Time Complexity:
 * - Insert/Delete: O(log n)
 * - Peek: O(1)
 * - Build: O(n)
 * 
 * JavaScript-Specific Notes:
 * - No built-in PriorityQueue (need custom implementation)
 * - Array.sort() is O(n log n) - avoid in loops
 * - Use custom heap class for better performance
 * - Consider using libraries for production code
 * 
 * This guide should help you quickly identify heap problems and choose the right approach!
 */

/**
 * This is a reference guide for heap problems.
 * No implementation needed - just concepts and patterns.
 */

function main() {
    console.log("=== HEAP DATA STRUCTURE GUIDE ===");
    console.log("Use this as a quick reference for heap problems.");
    console.log("Key patterns to recognize:");
    console.log("1. 'Top K' problems");
    console.log("2. 'Kth largest/smallest' problems");
    console.log("3. 'Stream' or 'real-time' problems");
    console.log("4. 'Scheduling with priority' problems");
    console.log("5. 'Frequent elements' problems");
    console.log("6. 'Merge sorted arrays/lists' problems");
    console.log();
    console.log("Choose heap type based on problem:");
    console.log("- Min Heap: K smallest, scheduling");
    console.log("- Max Heap: K largest, frequency");
    console.log("- Two Heaps: Median problems");
    console.log();
    console.log("JavaScript Implementation Tips:");
    console.log("- Use custom MinHeap class for better performance");
    console.log("- Avoid array.sort() in loops");
    console.log("- Consider heap size (K vs all elements)");
}

// Export for use in other modules
if (typeof module !== 'undefined' && module.exports) {
    module.exports = {
        main,
        // Add any helper functions here if needed
    };
} else {
    // Run if executed directly
    main();
} 